File: minecraftmod/platform-src/minecraft/mc-26.2/src/main/clojure/cn/li/mc262/datagen/item_registry.clj
(ns cn.li.mc262.datagen.item-registry
  "Shared item registry safety helpers for datagen (26.2 Identifier/getValue)."
  (:import [java.util HashSet]
           [net.minecraft.core.registries BuiltInRegistries]
           [net.minecraft.resources Identifier]
           [net.minecraft.world.item Items]))

(defn known-item-ids
  [get-item-ids-fn get-item-name-fn get-block-ids-fn get-block-name-fn mod-id]
  (let [known (HashSet.)]
    (doseq [item-id (get-item-ids-fn)]
      (let [registry-name (get-item-name-fn item-id)]
        (.add known (str mod-id ":" registry-name))))
    (doseq [block-id (get-block-ids-fn)]
      (let [registry-name (get-block-name-fn block-id)]
        (.add known (str mod-id ":" registry-name))))
    known))

(defn item-exists?
  [^String id known parse-rl-fn]
  (or (.contains ^HashSet known id)
      (let [^Identifier rl (parse-rl-fn id)
            item (.getValue BuiltInRegistries/ITEM rl)]
        (and (some? item)
             (not= item Items/AIR)
             (not= "minecraft:air" (str (.getKey BuiltInRegistries/ITEM item)))))))

(defn safe-item-id
  [id known parse-rl-fn]
  (let [sid (str id)]
    (if (item-exists? sid known parse-rl-fn)
      sid
      "minecraft:book")))

(defn with-safe-items
  [criteria known parse-rl-fn]
  (mapv
   (fn [entry]
     (if (= :inventory-changed (:type entry))
       (update entry :items (fn [xs] (mapv #(safe-item-id % known parse-rl-fn) xs)))
       entry))
   criteria))

File: minecraftmod/platform-src/minecraft/mc-26.2/src/main/java/cn/li/mc262/runtime/ItemRegistry.java
package cn.li.mc262.runtime;

import cn.li.mcbase.runtime.ItemPlayerOps;

import cn.li.mcver.ResourceLocations;
import net.minecraft.core.registries.BuiltInRegistries;
import net.minecraft.resources.Identifier;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.Items;

public final class ItemRegistry {
    static {
        ItemPlayerOps.installItemIdLookup(ItemRegistry::getItemById);
    }

    private ItemRegistry() {}

    public static Item get(Identifier id) {
        return BuiltInRegistries.ITEM.getValue(id);
    }

    public static Item getItemById(String itemId) {
        try {
            if (itemId == null || itemId.isEmpty()) return null;
            Item item = BuiltInRegistries.ITEM.getValue(ResourceLocations.parse(itemId));
            return item == Items.AIR ? null : item;
        } catch (Exception ignored) {
            return null;
        }
    }

    public static ItemStack createItemStackById(String itemId, int count) {
        Item item = getItemById(itemId);
        if (item == null || count <= 0) {
            return ItemStack.EMPTY;
        }
        return new ItemStack(item, count);
    }

    public static String getItemKey(Item item) {
        Identifier id = BuiltInRegistries.ITEM.getKey(item);
        return id != null ? id.toString() : null;
    }
}
